homeassistant/components/zimi/light.py

```diff
@@ -0,0 +1,136 @@
+
+    api: ControlPoint = config_entry.runtime_data
+
+    entities = [ZimiLight(device, api) for device in api.lights]
+
+    async_add_entities(entities)
+
+
+class ZimiLight(LightEntity):
+    """Representation of a Zimi Light."""
+
+    def __init__(self, light: ControlPointDevice, api: ControlPoint) -> None:
+        """Initialize an ZimiLight."""
+
+        self._attr_unique_id = light.identifier
+        self._attr_should_poll = False
+        self._light = light
+        self._light.subscribe(self)
+        self._state = False
+        self._brightness = None
+        if self._light.type == "dimmer":
+            self._attr_color_mode = ColorMode.BRIGHTNESS
+            self._attr_supported_color_modes = {ColorMode.BRIGHTNESS}
+        else:
+            self._attr_color_mode = ColorMode.ONOFF
+            self._attr_supported_color_modes = {ColorMode.ONOFF}
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, light.identifier)},
+            name=self._light.name,
+            suggested_area=self._light.room,
+            via_device=(DOMAIN, api.host + ":" + str(api.port)),
+        )
+        self.update()
+        _LOGGER.debug("Initialising %s in %s", self.name, self._light.room)
+
+    def __del__(self):
+        """Cleanup ZimiLight with removal of notification."""
+        self._light.unsubscribe(self)
+
+    @property
+    def name(self) -> str:
+        """Return the display name of this light."""
+        return self._name.strip()
+
+    @property
+    def available(self) -> bool:
+        """Return True if Home Assistant is able to read the state and control the underlying device."""
+        return self._light.is_connected
+
+    @property
+    def brightness(self) -> int | None:
+        """Return the brightness of the light.
+
+        This method is optional. Removing it indicates to Home Assistant
+        that brightness is not supported for this light.
+        """
+        return self._brightness
+
+    @property
+    def is_on(self) -> bool:
+        """Return true if light is on."""
+        return self._state
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Instruct the light to turn on.
+
+        You can skip the brightness part if your light does not support
+        brightness control.
+        """
+
+        _LOGGER.debug(
+            "Sending turn_on(brightness=%d) for %s",
+            kwargs.get(ATTR_BRIGHTNESS, 255) * 100 / 255,
+            self.name,
+        )
+
+        if self._light.type == "dimmer":
+            await self._light.set_brightness(
+                kwargs.get(ATTR_BRIGHTNESS, 255) * 100 / 255
+            )
+        else:
+            await self._light.turn_on()
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Instruct the light to turn off."""
+
+        _LOGGER.debug("Sending turn_off() for %s", self.name)
+
+        await self._light.turn_off()
+
+    def notify(self, _observable):
```

Comment:
I can't seem to find where is this passed to/called from?

Reply:
During __init__ I call self._light.subscribe(self) to subscribe to the notification queu on the device.   

Within the zcc library you will see in the device.py file I define subscribe() and unsubscribe() methods with add/del observers as well as utility methods to notify all observers when an event occurs.   The notify includes a reference back to the notifying device which is ignored as I just trigger HA to schedule and update() on this device.